File: public/arcade/star-crash/crash.py
import math
import random
from typing import Optional

from js import document, window
from pyodide.ffi import create_proxy

import crash_ui

STATE = {
  "balance": 300,
  "bet": 50,
  "running": False,
  "cashed_out": False,
  "multiplier": 1.0,
  "crash_point": 0.0,
  "start_time": 0.0,
  "animation_proxy": None,
  "history": [],
  "round_index": 1,
  "last_cashout": None
}

EVENT_PROXIES = []

canvas = crash_ui.elements["canvas"]
CTX = canvas.getContext("2d")
canvas_width = 0
canvas_height = 0


def resize_canvas() -> None:
  global canvas_width, canvas_height
  rect = canvas.getBoundingClientRect()
  canvas.width = math.floor(rect.width * window.devicePixelRatio)
  canvas.height = math.floor(rect.height * window.devicePixelRatio)
  canvas.style.width = f"{rect.width}px"
  canvas.style.height = f"{rect.height}px"
  canvas_width = canvas.width
  canvas_height = canvas.height
  draw_chart(STATE["multiplier"])


def format_multiplier(value: float) -> str:
  return f"x{value:.2f}"


def format_stars(value: int) -> str:
  return f"{value:,}".replace(",", " ")


def update_balance_display() -> None:
  crash_ui.elements["balance"].textContent = f"{format_stars(STATE['balance'])} ★"


def update_bet_display() -> None:
  crash_ui.elements["betDisplay"].textContent = f"{STATE['bet']} ★"


def set_status(text: str, status: Optional[str] = None) -> None:
  crash_ui.elements["statusBadge"].textContent = text
  badge = crash_ui.elements["statusBadge"]
  badge.classList.remove("status-live", "status-prep", "status-crash", "status-win")
  if status:
    badge.classList.add(f"status-{status}")


def random_crash_point() -> float:
  roll = random.random()
  crash_value = 1 / (1 - roll)
  return round(min(max(crash_value, 1.2), 15.0), 2)


def draw_chart(multiplier: float) -> None:
  CTX.save()
  CTX.clearRect(0, 0, canvas_width, canvas_height)

  CTX.scale(window.devicePixelRatio, window.devicePixelRatio)

  width = canvas_width / window.devicePixelRatio
  height = canvas_height / window.devicePixelRatio

  CTX.fillStyle = "rgba(12,16,28,0.65)"
  CTX.fillRect(0, 0, width, height)

  CTX.strokeStyle = "rgba(255,255,255,0.12)"
  CTX.lineWidth = 1
  CTX.beginPath()
  for i in range(6):
    y = height - (i * (height / 5)) - 24
    CTX.moveTo(48, y)
    CTX.lineTo(width - 24, y)
  CTX.stroke()

  CTX.strokeStyle = "rgba(59,130,246,0.9)"
  CTX.lineWidth = 2.4
  CTX.beginPath()
  CTX.moveTo(48, height - 48)

  max_multiplier = max(multiplier, 2.0)
  segments = 60
  for step in range(1, segments + 1):
    progress = step / segments
    x = 48 + progress * (width - 72)
    y_multiplier = max_multiplier * progress
    y = height - 48 - (y_multiplier * 20)
    if y < 48:
      y = 48
    CTX.lineTo(x, y)
  CTX.stroke()

  CTX.restore()


def update_multiplier_display(value: float) -> None:
  crash_ui.elements["multiplierLabel"].textContent = format_multiplier(value)
  draw_chart(value)


def add_history_entry(result: str, payout: int) -> None:
  feed = crash_ui.elements["history"]
  placeholder = feed.querySelector(".history-placeholder")
  if placeholder:
    feed.removeChild(placeholder)

  card = document.createElement("div")
  card.className = "history-entry"

  title = document.createElement("strong")
  title.textContent = f"Раунд #{STATE['round_index']}"
  card.appendChild(title)

  status = document.createElement("span")
  status.style.fontSize = "12px"
  status.style.color = "rgba(244,244,245,0.65)"
  status.textContent = result
  card.appendChild(status)

  if payout:
    win = document.createElement("span")
    win.style.color = "rgba(16,185,129,0.9)"
    win.style.fontSize = "12px"
    win.textContent = f"Выплата: {format_stars(payout)} ★"
    card.appendChild(win)

  feed.prepend(card)
  STATE["history"].append({"result": result, "payout": payout})
  if len(STATE["history"]) > 6:
    STATE["history"].pop(0)
    last = feed.lastElementChild
    if last:
      feed.removeChild(last)


def end_round(crashed: bool) -> None:
  STATE["running"] = False
  crash_ui.buttons["start"].disabled = False
  crash_ui.buttons["cashout"].disabled = True
  STATE["round_index"] += 1
  if crashed:
    set_status("Краш", "crash")
  else:
    set_status("Ожидание", "prep")


def animation_frame(timestamp):
  if not STATE["running"]:
    return

  elapsed = (timestamp - STATE["start_time"]) / 1000
  growth = 1.0 + elapsed * 1.4
  multiplier = min(growth, STATE["crash_point"])
  STATE["multiplier"] = multiplier
  update_multiplier_display(multiplier)

  if multiplier >= STATE["crash_point"]:
    crash_ui.showToast("Crash! Корабль обрушился.")
    add_history_entry(f"Crash на {format_multiplier(STATE['crash_point'])}", 0)
    end_round(True)
    return

  STATE["animation_proxy"] = window.requestAnimationFrame(ANIMATION_PROXY)


def start_round(_event=None) -> None:
  if STATE["running"]:
    return

  bet_input = crash_ui.elements["betInput"]
  try:
    bet_value = int(bet_input.value)
  except Exception:
    bet_value = STATE["bet"]

  bet_value = max(1, min(500, bet_value))

  if bet_value > STATE["balance"]:
    crash_ui.showToast("Недостаточно звёзд для ставки.")
    return

  STATE["bet"] = bet_value
  crash_ui.elements["betInput"].value = str(bet_value)
  update_bet_display()

  STATE["balance"] -= bet_value
  update_balance_display()

  STATE["running"] = True
  STATE["cashed_out"] = False
  STATE["multiplier"] = 1.0
  STATE["crash_point"] = random_crash_point()
  STATE["start_time"] = window.performance.now()

  crash_ui.buttons["start"].disabled = True
  crash_ui.buttons["cashout"].disabled = False
  set_status("В полёте", "live")
  crash_ui.showToast("Раунд запущен!")

  update_multiplier_display(1.0)
  STATE["animation_proxy"] = window.requestAnimationFrame(ANIMATION_PROXY)


def cashout(_event=None) -> None:
  if not STATE["running"] or STATE["cashed_out"]:
    return

  payout = int(STATE["bet"] * STATE["multiplier"])
  STATE["balance"] += payout
  update_balance_display()
  crash_ui.showToast(f"Вы забрали {format_stars(payout)} ★!")
  add_history_entry(f"Cashout на {format_multiplier(STATE['multiplier'])}", payout)
  STATE["cashed_out"] = True
  end_round(False)


def apply_chip(event) -> None:
  value = event.currentTarget.dataset.bet
  if not value:
    return
  crash_ui.elements["betInput"].value = value
  STATE["bet"] = int(value)
  update_bet_display()


def sync_bet_input(_event=None) -> None:
  try:
    value = int(crash_ui.elements["betInput"].value)
  except Exception:
    value = STATE["bet"]
  value = max(1, min(500, value))
  crash_ui.elements["betInput"].value = str(value)
  STATE["bet"] = value
  update_bet_display()


def clear_history(_event=None) -> None:
  STATE["history"].clear()
  feed = crash_ui.elements["history"]
  feed.innerHTML = ""
  placeholder = document.createElement("p")
  placeholder.className = "history-placeholder"
  placeholder.textContent = "Запустите раунд, чтобы увидеть результаты."
  feed.appendChild(placeholder)
  crash_ui.showToast("История очищена")


def setup() -> None:
  resize_canvas()
  update_balance_display()
  update_bet_display()
  set_status("Ожидание", "prep")

  start_proxy = create_proxy(start_round)
  cashout_proxy = create_proxy(cashout)
  clear_proxy = create_proxy(clear_history)
  input_proxy = create_proxy(sync_bet_input)
  resize_proxy = create_proxy(lambda _event=None: resize_canvas())

  EVENT_PROXIES.extend([start_proxy, cashout_proxy, clear_proxy, input_proxy, resize_proxy])

  crash_ui.buttons["start"].addEventListener("click", start_proxy)
  crash_ui.buttons["cashout"].addEventListener("click", cashout_proxy)
  crash_ui.buttons["clearHistory"].addEventListener("click", clear_proxy)
  crash_ui.elements["betInput"].addEventListener("change", input_proxy)
  window.addEventListener("resize", resize_proxy)

  for chip in crash_ui.buttons["chips"]:
    proxy = create_proxy(apply_chip)
    EVENT_PROXIES.append(proxy)
    chip.addEventListener("click", proxy)


ANIMATION_PROXY = create_proxy(animation_frame)

setup()

File: public/arcade/star-crash/crash.js
const root = document.getElementById('crash-root');
const toast = document.getElementById('crash-toast');
const toastMessage = document.getElementById('crash-toast-message');

const elements = {
  balance: document.getElementById('crash-balance'),
  betDisplay: document.getElementById('crash-bet-display'),
  multiplierLabel: document.getElementById('multiplier-label'),
  statusBadge: document.getElementById('round-status'),
  history: document.getElementById('crash-history'),
  canvas: document.getElementById('crash-canvas'),
  betInput: document.getElementById('bet-input')
};

const buttons = {
  start: document.getElementById('btn-start-round'),
  cashout: document.getElementById('btn-cashout'),
  clearHistory: document.getElementById('btn-clear-history'),
  chips: document.querySelectorAll('.chip-btn')
};

const defaultConfig = {
  baseBet: 50,
  maxMultiplier: 12,
  autoCashout: 0,
  roundDelay: 4
};

const state = {
  config: { ...defaultConfig },
  balance: 300,
  bet: 50,
  running: false,
  awaitingStart: false,
  multiplier: 1,
  crashPoint: 2.4,
  history: [],
  animationFrame: null,
  startTime: 0,
  elapsed: 0,
  serverSynced: false,
  initDataRaw: null,
  balancePollId: null,
  path: []
};

function showToast(message) {
  toastMessage.textContent = message;
  toast.classList.add('is-visible');
  clearTimeout(showToast.timer);
  showToast.timer = setTimeout(() => {
    toast.classList.remove('is-visible');
  }, 2200);
}

function formatMultiplier(value) {
  return `x${value.toFixed(2)}`;
}

function updateBalanceDisplay() {
  elements.balance.textContent = `${Math.max(0, Math.round(state.balance))} ★`;
}

async function fetchBalance() {
  if (!state.initDataRaw) {
    return;
  }
  try {
    const response = await fetch('/api/mini-app/balance', {
      headers: {
        Accept: 'application/json',
        'x-telegram-init-data': state.initDataRaw
      }
    });
    if (!response.ok) {
      return;
    }
    const payload = await response.json();
    if (payload?.balance?.available !== undefined) {
      state.balance = Math.max(0, Math.round(payload.balance.available));
      state.serverSynced = true;
      updateBalanceDisplay();
    }
  } catch (error) {
    // ignore
  }
}

function startBalancePolling() {
  if (state.balancePollId) {
    clearInterval(state.balancePollId);
  }
  state.balancePollId = setInterval(fetchBalance, 15000);
}

function updateBetDisplay() {
  elements.betDisplay.textContent = `${Math.round(state.bet)} ★`;
  elements.betInput.value = Math.round(state.bet);
}

function setStatus(label, variant) {
  elements.statusBadge.textContent = label;
  elements.statusBadge.dataset.variant = variant || '';
}

function updateMultiplierDisplay() {
  elements.multiplierLabel.textContent = formatMultiplier(state.multiplier);
}

function setControlsEnabled(isEnabled) {
  buttons.start.disabled = !isEnabled;
  buttons.cashout.disabled = isEnabled;
  elements.betInput.disabled = !isEnabled;
  buttons.chips.forEach((button) => {
    button.disabled = !isEnabled;
  });
}

function updateHistory() {
  if (!state.history.length) {
    elements.history.innerHTML =
      '<p class="history-placeholder">Запустите раунд, чтобы увидеть результаты.</p>';
    return;
  }

  elements.history.innerHTML = '';
  state.history.slice(-8).reverse().forEach((entry) => {
    const row = document.createElement('div');
    row.className = 'history-row';
    row.innerHTML = `
      <span>${formatMultiplier(entry.crash)}</span>
      <span>${entry.result}</span>
      <span>${entry.payout > 0 ? `+${entry.payout} ★` : '—'}</span>
    `;
    elements.history.appendChild(row);
  });
}

const ctx = elements.canvas.getContext('2d');

function resizeCanvas() {
  const { width, height } = elements.canvas.getBoundingClientRect();
  const dpr = window.devicePixelRatio || 1;
  elements.canvas.width = Math.round(width * dpr);
  elements.canvas.height = Math.round(height * dpr);
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
}

function drawChart() {
  const { width, height } = elements.canvas.getBoundingClientRect();
  ctx.clearRect(0, 0, width, height);
  ctx.fillStyle = 'rgba(255,255,255,0.04)';
  ctx.fillRect(0, 0, width, height);

  ctx.strokeStyle = 'rgba(255, 214, 10, 0.9)';
  ctx.lineWidth = 2;
  ctx.beginPath();
  const points = state.path;
  if (!points.length) {
    ctx.stroke();
    return;
  }
  points.forEach((point, index) => {
    const x = (point.t / 12) * width;
    const y = height - (point.mult / state.config.maxMultiplier) * (height - 12) - 6;
    if (index === 0) {
      ctx.moveTo(x, y);
    } else {
      ctx.lineTo(x, y);
    }
  });
  ctx.stroke();
}

function generateCrashPoint(maxMultiplier) {
  const roll = Math.random();
  const weighted = 1 / Math.max(0.08, roll);
  return Math.min(maxMultiplier, Math.max(1.1, weighted));
}

function stopAnimation() {
  if (state.animationFrame) {
    cancelAnimationFrame(state.animationFrame);
    state.animationFrame = null;
  }
}

function finalizeRound(result, payout) {
  state.history.push({
    crash: state.multiplier,
    result,
    payout
  });
  updateHistory();
  setControlsEnabled(true);
  state.running = false;
  state.awaitingStart = false;
  state.elapsed = 0;
  state.path = [];
  updateBalanceDisplay();
  setStatus('Ожидание', 'idle');
}

function crashRound() {
  stopAnimation();
  setStatus('Обрушение', 'crash');
  showToast('Множитель обрушился!');
  finalizeRound('Crash', 0);
  updateMultiplierDisplay();
  drawChart();
}

function tickFrame(timestamp) {
  if (!state.running) {
    return;
  }
  if (!state.startTime) {
    state.startTime = timestamp;
  }
  const elapsedSeconds = (timestamp - state.startTime) / 1000;
  state.elapsed = elapsedSeconds;
  const growth = 1 + elapsedSeconds * 0.7 + Math.pow(elapsedSeconds, 1.3) * 0.22;
  state.multiplier = Math.min(state.config.maxMultiplier, growth);
  state.path.push({ t: elapsedSeconds, mult: state.multiplier });
  updateMultiplierDisplay();
  drawChart();

  if (state.multiplier >= state.crashPoint) {
    crashRound();
    return;
  }

  if (state.config.autoCashout && state.multiplier >= state.config.autoCashout) {
    handleCashout();
    return;
  }

  state.animationFrame = requestAnimationFrame(tickFrame);
}

function handleStart() {
  if (state.running || state.awaitingStart) {
    showToast('Раунд уже готовится');
    return;
  }
  if (state.bet <= 0 || state.bet > state.balance) {
    showToast('Недостаточно звёзд для ставки');
    return;
  }

  if (!state.serverSynced) {
    state.balance -= state.bet;
  }
  updateBalanceDisplay();
  setControlsEnabled(false);
  setStatus('Подготовка', 'waiting');
  state.awaitingStart = true;
  showToast('Готовим запуск...');

  setTimeout(() => {
    state.awaitingStart = false;
    state.running = true;
    state.multiplier = 1;
    state.crashPoint = generateCrashPoint(state.config.maxMultiplier);
    state.path = [];
    state.startTime = 0;
    setStatus('Режим x', 'running');
    state.animationFrame = requestAnimationFrame(tickFrame);
  }, Math.max(1, state.config.roundDelay) * 1000);
}

function handleCashout() {
  if (!state.running) {
    showToast('Раунд ещё не запущен');
    return;
  }
  stopAnimation();
  const payout = Math.max(0, Math.round(state.bet * state.multiplier));
  if (!state.serverSynced) {
    state.balance += payout;
  }
  showToast(`Вы забрали ${payout} ★`);
  finalizeRound('Cashout', payout);
  updateMultiplierDisplay();
  drawChart();
}

function setBet(value) {
  const sanitized = Math.max(1, Math.round(value));
  state.bet = sanitized;
  updateBetDisplay();
}

function applyConfig(nextConfig) {
  state.config = { ...state.config, ...nextConfig };
  if (state.config.baseBet) {
    setBet(state.config.baseBet);
  }
  updateMultiplierDisplay();
}

buttons.start.addEventListener('click', handleStart);
buttons.cashout.addEventListener('click', handleCashout);
buttons.clearHistory.addEventListener('click', () => {
  state.history = [];
  updateHistory();
});
elements.betInput.addEventListener('change', (event) => {
  const value = Number.parseInt(event.target.value, 10);
  if (Number.isFinite(value)) {
    setBet(value);
  }
});
buttons.chips.forEach((button) => {
  button.addEventListener('click', () => {
    const value = Number.parseInt(button.dataset.bet || '0', 10);
    if (value) {
      setBet(value);
    }
  });
});

window.addEventListener('message', (event) => {
  const data = event.data || {};
  if (data.type === 'STAR_CRASH_CONFIG' && data.payload) {
    applyConfig(data.payload);
  }
  if (data.type === 'STAR_CRASH_BALANCE' && typeof data.payload?.available === 'number') {
    state.balance = Math.max(0, Math.round(data.payload.available));
    state.serverSynced = true;
    updateBalanceDisplay();
  }
  if (data.type === 'STAR_CRASH_AUTH' && data.payload?.initDataRaw) {
    state.initDataRaw = data.payload.initDataRaw;
    fetchBalance();
    startBalancePolling();
  }
});

resizeCanvas();
updateBalanceDisplay();
updateBetDisplay();
updateMultiplierDisplay();
updateHistory();
setStatus('Ожидание', 'idle');

window.addEventListener('resize', () => {
  resizeCanvas();
  drawChart();
});
